### ISSUE TITLE
macOS | Requirements too strict?

### ISSUE THREAD

**Issue by tmontes** [beeware/briefcase#270]
## Context

Hi BeeWare,

I've been contributing to the [Mu Editor](https://github.com/mu-editor/mu/) and, in the context of [this Mu macOS packaging issue](https://github.com/mu-editor/mu/issues/925) came to find out that a newer `briefcase` version `0.2.10` has been released -- it wasn't so obvious because, up until now (and until further discussion), we've locked macOS packaging dependencies on `briefcase==0.2.9`.

To address the fact that the `briefcase`-produced *application bundles* do not include the underlying project version, I just [raised a PR](https://github.com/beeware/Python-macOS-template/pull/3) against the 3.6 branch of [Python-macOS-template](https://github.com/beeware/Python-macOS-template/). :)

## Surprise

While exploring the inner workings of `briefcase` and its interaction with `Python-macOS-template`, I brought up my Mu development environment to the latest `briefcase` version, only to be surprised by the following message:

```
 * Confirming prerequisites...

 * Looking for Xcode...
xcode-select: error: tool 'xcodebuild' requires Xcode, but active developer directory '/Library/Developer/CommandLineTools' is a command line tools instance
Briefcase requires Xcode, but it does not appear to be installed.
Please install the latest version of Xcode using the Mac App Store:

    https://itunes.apple.com/us/app/xcode/id497799835
```

## Details

I digged a bit and found that such check was merged in #227 by @amymok who wrote:

> If Xcode and Command Line Tools are not installed on the mac, an DMGError would triggered when building MacOS Application

Trying to avoid installing XCode on my system -- I never really needed it, having only the *XCode Command Line Tools* installed -- I commented out the call to `confirm_prerequisites()` and took `briefcase` for a spin on my system, packaging Mu. Along with my expectations, everything worked nicely... Here are the combinations I tried:

* Plain `python setup.py macos` - worked ok.
* Build a DMG `python setup.py macos --build` -- worked ok.
* Using a custom support package with `python setup.py macos --support-pkg=https://github.com/mu-editor/mu_portable_python_macos/releases/download/0.0.6/python3-reduced.tar.gz` -- this is how Mu is packaged -- worked ok.
* Using a custom support package and producing a DMG with `python setup.py macos --support-pkg=https://github.com/mu-editor/mu_portable_python_macos/releases/download/0.0.6/python3-reduced.tar.gz --build` -- an extension of how Mu is packaged -- worked ok.

About my system:
* macOS Sierra 10.12.6.
* `gcc --version`:

    ```
    Configured with: --prefix=/Library/Developer/CommandLineTools/usr --with-gxx-include-dir=/usr/include/c++/4.2.1
    Apple LLVM version 9.0.0 (clang-900.0.39.2)
    Target: x86_64-apple-darwin16.7.0
    Thread model: posix
    InstalledDir: /Library/Developer/CommandLineTools/usr/bin
    ```
* `clang --version`:

    ```
    Apple LLVM version 9.0.0 (clang-900.0.39.2)
    Target: x86_64-apple-darwin16.7.0
    Thread model: posix
    InstalledDir: /Library/Developer/CommandLineTools/usr/bin
    ```

As a final note about my system, and Python environment in particular, I notice that `dmgbuild==1.3.2` is installed, pretty much as specified in `briefcase`'s requirements.

## Question / Suggestion

With all of this said, I'm not really sure about the particulars that prompted the merging of such apparently too-strict requirements to package macOS applications. Could it be that on other macOS versions `dmgbuild` does require a full XCode install? Or was the requirement put in place over-zealously? What about users that don't want/need to create DMG files? Should they be required to install XCode if they don't need it? Hmmm... Lots of questions! :)

I would like to understand @amymok motivations and criteria for #227 and, hopefully, determine that a full XCode installation is not needed and, thus, propose a PR to drop that requirement. Not wanting to do that from an uninformed position, I decided to create this issue.

Thanks again for `briefcase` (and BeeWare!) and, in advance, for any feedback and insights into this topic. :)

**Comment by tmontes** [comment]
## Addendum

I investigated this further:
* Reviewed both `src/briefcase/app.py` and `src/briefcase/macos.py`.
* Reviewed the source for [`dmgbuild`](https://github.com/al45tair/dmgbuild).

Summary:
* Neither one seems to depend on XCode or any of the other tools.
* The `DMGError` exceptions noted by @amymok in #227 are raised by `dmgbuild` in its [`dmgbuild/core.py`](https://github.com/al45tair/dmgbuild/blob/master/dmgbuild/core.py) module -- unfortunately, no further context or detail was shared then.
* All such `raise DMGError`s follow unsuccessful calls to the [`hdiutil` function here](https://github.com/al45tair/dmgbuild/blob/master/dmgbuild/core.py#L45), which essentially runs the `/usr/bin/hdiutil` macOS binary in a sub-process.
* AFAICT, the `/usr/bin/hdiutil` binary is included in all macOS installations, regardless of whether XCode or its Command Line Tools are installed or not.

Statement:
Thus, I see no reason for the whole of `confirm_prerequisites` in `src/briefcase/macos.py`: XCode not required, GCC not required, XCode Command Line Tools not required. I am still perfectly open to be shown otherwise, though...

PS: Wondering what might actually have been @amymok's experience: given that `/usr/bin/hdiutil` is essentially an I/O tool and that I/O by itself can fail for all sorts of reasons, it might be that whichever failing condition just "went away by itself" after XCode was installed in a somehow misguided (?) "fix attempt". Like I said, wondering... :(


Would love some feedback on this issue.
Thanks once again.

**Comment by freakboy3742** [comment]
Thanks for the investigation work on this. 

FWIW, I can't recall the exact errors that lead to this change being added to macOS specifically. iOS obviously has a hard XCode requirement, and I have a vague recollection of problems existing that were resolved by installing XCode - but I also can't argue with the results of your investigation work. 

It's entirely possible this is actually an accidental merge into the wrong backend. It looks like iOS *doesn't* have the iOS checking, when it probably should, and I merged the original patch in haste during the PyCon US sprints...

**Comment by tmontes** [comment]
> It's entirely possible this is actually an accidental merge into the wrong backend. It looks like iOS _doesn't_ have the iOS checking, when it probably should, and I merged the original patch in haste during the PyCon US sprints...

I wondered about that possibility, too. :)

**Comment by freakboy3742** [comment]
This should have been addressed with the 0.3 development version that was just merged to master. Addressing #174 may involve reintroducing *some* of these requirements.

@tmontes As an aside, I'd be interested in any feedback you have on changes in 0.3 branch. It's a radical change to the configuration format; submitting a patch to Mu to update to the new format is on my todo list (since Mu is one of our known uses in the wild!)
